bootstrap: correctly handle doc paths within submodules

jieyouxu opened the pull request:
Fixes #135041 by passing the correct submodule path when requiring submodules. This PR changes `is_path_in_submodule` to `submodule_path_of`. `submodule_path_of` returns the path of the containing submodule when given a path nested inside a submodule we handle, and `None` otherwise.

I tested this manually locally by unregistering the `src/tools/cargo` submodule, then running `./x doc src/tools/cargo/src/doc`. This command fails on master with

```
thread 'main' panicked at src/bootstrap/src/utils/helpers.rs:441:5:
std::fs::read_dir(dir) failed with No such file or directory (os error 2)
```

since the require submodule fails as `src/tools/cargo/src/doc` is not a known submodule. Now we use the submodule path if such a nested-in-submodule-path is passed, and thus running this command with cargo submodule unregistered still succeeds:

```
Rustbook (x86_64-unknown-linux-gnu) - cargo
Doc path: /home/joe/repos/rust/build/x86_64-unknown-linux-gnu/doc/cargo/index.html
Build completed successfully in 0:00:11
```

r? @onur-ozkan

jieyouxu commented:
Not *super* happy with how this macro still looks (it's still a bit convoluted), but at least the invocation-site is a bit nicer to use, I suppose.

There's also some pre-existing issues, such as:

- We do not really sanity-check the doc path and submodule path, e.g. that they share prefix. That if the doc is inside a submodule, that the `Step` author correctly passed a path where (1) it's a path of a submodule that we handle and (2) it's the *same* submodule as where the doc resides in.
- Not sure why `requires_submodule` silently fails? I think we just don't check it maybe?

jieyouxu commented on src/bootstrap/src/core/build_steps/doc.rs:
```diff
@@ -34,49 +53,59 @@ macro_rules! book {
 // NOTE: When adding a book here, make sure to ALSO build the book by
 // adding a build step in `src/bootstrap/code/builder/mod.rs`!
 // NOTE: Make sure to add the corresponding submodule when adding a new book.
-// FIXME: Make checking for a submodule automatic somehow (maybe by having a list of all submodules
-// and checking against it?).
```
I think this is actually addressed by #134967.

jieyouxu commented:
@onur-ozkan do you think we should fix-forward (i.e. this PR) or revert #134967 then try to reland that?

onur-ozkan commented:
> @onur-ozkan do you think we should fix-forward (i.e. this PR) or revert #134967 then try to reland that?

I think we can fix it right away. I need couple minutes to review this.

onur-ozkan commented on src/bootstrap/src/core/build_steps/doc.rs:
```diff
@@ -20,10 +20,29 @@ use crate::core::builder::{
+/// need to use the submodule path to checkout the submodule. The doc could be inside an submodule
+/// that is not checked out. Helper for [`book`].
+fn select_book_submodule_path<'a>(doc_path: &'a str, submodule_path: Option<&'a str>) -> &'a str {
+    submodule_path.unwrap_or(doc_path)
+}
+
 macro_rules! book {
-    ($($name:ident, $path:expr, $book_name:expr, $lang:expr ;)+) => {
-        $(
-            #[derive(Debug, Clone, Hash, PartialEq, Eq)]
+    (
+        $name:ident {
+            name: $book_name:expr,
+            // This can be *within* a submodule, e.g. `src/tools/cargo/src/doc`, compared to the
+            // submodule *itself*.
+            doc_path: $doc_path:expr
+            // Optional; only needed if `doc_path` is inside the submodule path, i.e. they are not
+            // the same.
+            $(, submodule_path: $submodule_path:expr )?
+            // Optional; only needed if the book has multi-lingual support.
+            $(, languages: $languages:expr )?
+            $( , )?
+        }
+    ) => {
+        #[derive(Debug, Clone, Hash, PartialEq, Eq)]
```
Please correct me if I am wrong: the problem is that we are trying to update a submodule that is actually not a submodule but a subdirectory within a submodule. If that's the case, there is no need to overcomplicate this macro again (which kills the purpose of https://github.com/rust-lang/rust/pull/134967). Instead, make [is_path_in_submodule](https://github.com/onur-ozkan/rust/blob/6eb9ebf1d61131df43f18e379493e23a193c3a7c/src/bootstrap/src/utils/helpers.rs#L63-L64) to return the submodule path as `Some(path)` instead of returning `true` and return `None` instead of `false`.

jieyouxu replied:
Sounds reasonable, let me do that a bit later.

jieyouxu replied:
Changed `is_path_in_submodule` to `submodule_path_of`, which returns `Some(submodule_path)` if the destination is inside a submodule, and return `None` if it's not a submodule we handle.

onur-ozkan replied:
Thanks

onur-ozkan commented:
@rustbot author

jieyouxu committed:
bootstrap: make `is_path_in_submodule` return optional submodule path instead

jieyouxu commented:
Changes since last review:

- Dropped unrelated macro changes.
- Rename `is_path_in_submodule` to `submodule_path_of`. `submodule_path_of` returns `Some(submodule_path)` if the given path is nested inside a known submodule, and `None` otherwise.
- Adjusted the helper's tests as suitable.

@rustbot ready

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12611417884/job/35147058098) [(plain)](https://github.com/rust-lang/rust/commit/399e69eb6663745464229c6a97b66892d34eb240/checks/35147058098/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking home v0.5.9
    Checking termcolor v1.4.1
    Checking build_helper v0.1.0 (/checkout/src/build_helper)
    Checking xz2 v0.1.7
error: this `.into_iter()` call is equivalent to `.iter()` and will not consume the `slice`
   |
66 |     submodule_paths.into_iter().find_map(|submodule_path| {
   |                     ^^^^^^^^^ help: call directly: `iter`
   |
```

</details>

onur-ozkan reviewed (approved):
LGTM, feel free to r=me once CI goes green

jieyouxu commented:
PR CI is 🍏 
@bors r=@onur-ozkan rollup

bors commented:
:pushpin: Commit 4406f42a1851108d745d281449bdd7bbaae35da9 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 4406f42a1851108d745d281449bdd7bbaae35da9 -->
<!-- homu: {"type":"Approved","sha":"4406f42a1851108d745d281449bdd7bbaae35da9","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
@bors p=1 (contributor roadblock, feel free to include in rollup)

jieyouxu commented:
@bors p=6 (scheduling this before the next-in-line rollup)

bors commented:
:hourglass: Testing commit 4406f42a1851108d745d281449bdd7bbaae35da9 with merge ead4a8f53624f3b4a9ee833421fbf8a48bc07aca...
<!-- homu: {"type":"BuildStarted","head_sha":"4406f42a1851108d745d281449bdd7bbaae35da9","merge_sha":"ead4a8f53624f3b4a9ee833421fbf8a48bc07aca"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12612761658/job/35151919145)
Approved by: onur-ozkan
Pushing ead4a8f53624f3b4a9ee833421fbf8a48bc07aca to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"onur-ozkan","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12612761658/job/35151919145"},"merge_sha":"ead4a8f53624f3b4a9ee833421fbf8a48bc07aca"} -->

rust-timer commented:
Finished benchmarking commit (ead4a8f53624f3b4a9ee833421fbf8a48bc07aca): [comparison URL](https://perf.rust-lang.org/compare.html?start=2a8af4f7c8262a90b886bc063fe0c271d2b51a45&end=ead4a8f53624f3b4a9ee833421fbf8a48bc07aca&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=2a8af4f7c8262a90b886bc063fe0c271d2b51a45&end=ead4a8f53624f3b4a9ee833421fbf8a48bc07aca&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2a8af4f7c8262a90b886bc063fe0c271d2b51a45&end=ead4a8f53624f3b4a9ee833421fbf8a48bc07aca&stat=max-rss)
<details>
<summary>Results (secondary 0.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.8% | [0.8%, 0.8%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2a8af4f7c8262a90b886bc063fe0c271d2b51a45&end=ead4a8f53624f3b4a9ee833421fbf8a48bc07aca&stat=cycles:u)
<details>
<summary>Results (secondary 2.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.8% | [2.4%, 3.3%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=2a8af4f7c8262a90b886bc063fe0c271d2b51a45&end=ead4a8f53624f3b4a9ee833421fbf8a48bc07aca&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 762.575s -> 762.69s (0.02%)
**Artifact size**: 325.64 MiB -> 325.64 MiB (-0.00%)
